### PR TITLE
Migrating rest of integration tests to pytest

### PR DESCRIPTION
**Pull request by prady0t**
Addresses #4156

### CODE REVIEW

File: tests/integration/test_models/test_submodels/test_interface/test_butler_volmer.py
Comment (agriyakhetarpal) [inline tests/integration/test_models/test_submodels/test_interface/test_butler_volmer.py:12]: While it is fine here because we have both a setup method and a teardown method, I would suggest adding this item to the tracking issue as well – to migrate all of the `@pytest.fixture(autouse=True)` statements we've been using to the `pytest`-preferred style once all tests have been migrated. A single PR to do all of them would be great.

**Review by agriyakhetarpal — APPROVED** [review]

**Review by agriyakhetarpal — COMMENTED** [review]
I notice that some tests are failing just under the `macos-14` runner while they pass elsewhere, could you check that? The logs are here: https://github.com/pybamm-team/PyBaMM/actions/runs/10054485940/job/27789189858?pr=4285

File: tests/integration/test_models/test_submodels/test_interface/test_butler_volmer.py
Comment (prady0t) [reply tests/integration/test_models/test_submodels/test_interface/test_butler_volmer.py:12]: I can cover this in one PR after all the test migration. I'll just have to find and replace it would be difficult to track all of them down now and add them to tracking issue.

**Comment by prady0t** [comment]
> I notice that some tests are failing just under the `macos-14` runner while they pass elsewhere, could you check that? The logs are here: https://github.com/pybamm-team/PyBaMM/actions/runs/10054485940/job/27789189858?pr=4285

I'm not so sure. It's because of [this](https://numpy.org/doc/stable/reference/generated/numpy.testing.assert_allclose.html) function, but if we have an AssertionError it should be everywhere not just on macOS. It runs fine for me locally.

**Comment by agriyakhetarpal** [comment]
Sure, triggered a run by updating your branch

File: tests/integration/test_models/test_submodels/test_interface/test_butler_volmer.py
Comment (prady0t) [reply tests/integration/test_models/test_submodels/test_interface/test_butler_volmer.py:12]: Makes sense. Done!

**Comment by prady0t** [comment]
Strange it passes and fails sometimes.

**Comment by prady0t** [comment]
Could it be that our tests are non-deterministic since we are not using custom `Testcase` now?

**Comment by agriyakhetarpal** [comment]
Could you try stress-testing the integration test suite on macOS 14 locally, @prady0t?

**Comment by agriyakhetarpal** [comment]
> Could it be that our tests are non-deterministic since we are not using custom `Testcase` now?

It could be – maybe we should re-introduce the TestCase class for those tests until we are able to remove it completely.

**Comment by prady0t** [comment]
Let me stress test it then.

**Comment by prady0t** [comment]
I ran tests for a while but they are working fine 🤔

**Comment by agriyakhetarpal** [comment]
I've triggered another run, let's see how this goes

**Comment by agriyakhetarpal** [comment]
Discussed in the GSoC meeting on 24/07/2024 – restore the `TestCase` class so that the failing test classes derive from it temporarily until we merge #4231. Restoring that class should restore the determinism in the integration tests established in #2844.

**Comment by prady0t** [comment]
I've added back Testcase for failing test class.

**Comment by prady0t** [comment]
I think I accidentally changed the wrong file. I've changed it back could you re-run?

**Comment by prady0t** [comment]
It will still fail. I keep forgetting to test it locally 😞 . Let me try again

**Comment by agriyakhetarpal** [comment]
No worries – I cancelled the workflow here to redirect it to other branches/PRs and because it's going to fail here anyway, but GHA is completely free for open source code :)

**Review by Saransh-cpp — APPROVED** [review]

**Comment by agriyakhetarpal** [comment]
This PR has a few failures, but we have two approvals. Rather than dismissing any of the reviews, I'll mark this as a draft so that it doesn't get merged for now.

The current failure is:

```
Failed: Fixture "setUp" called directly. Fixtures are not meant to be called directly,
but are created automatically when test functions request them as parameters.
```

File: tests/integration/test_models/test_full_battery_models/test_lithium_ion/test_dfn.py
Comment (agriyakhetarpal) [inline tests/integration/test_models/test_full_battery_models/test_lithium_ion/test_dfn.py:39]:
```suggestion
class TestDFNWithSizeDistribution(TestCase):
```

**Review by agriyakhetarpal — APPROVED** [review]

**Review by arjxn-py — APPROVED** [review]
Thanks @prady0t, should be good to merge once @agriyakhetarpal's comments are resolved.

**Comment by prady0t** [comment]
We can merge this now once the CI pass

**Comment by agriyakhetarpal** [comment]
In the meantime CI passes, stress-testing this again a bunch of times locally

**Comment by agriyakhetarpal** [comment]
Could you fix the failing links here? https://github.com/pybamm-team/PyBaMM/actions/runs/10301504774/attempts/2#summary-28513370617

File: tests/integration/test_models/test_full_battery_models/test_lithium_ion/test_thermal_models.py
Comment (agriyakhetarpal) [inline tests/integration/test_models/test_full_battery_models/test_lithium_ion/test_thermal_models.py:102]: While it is okay here, I think the convention (whether in programming or in language) is to follow that errors connote a sense of negation and are hence measured as entities that are less than others, so this should have been the other way around. Let's make a note of this somewhere and fix all of them at once in another PR.

**Review by agriyakhetarpal — APPROVED** [review]

**Pull request merged**
